### PR TITLE
RUST-645 / RUST-1245 Add Atlas planned maintenance testing workload executor

### PR DESCRIPTION
**Pull request by isabelatkinson**
Workload executor specification: https://mongodb-labs.github.io/drivers-atlas-testing/spec-workload-executor.html

There are still a few kinks in the implementation that I'm working through in testing, but the bulk of the code should be ready for review.

### CODE REVIEW

File: src/test/atlas_planned_maintenance_testing/mod.rs
Comment (kmahar) [reply src/test/atlas_planned_maintenance_testing/mod.rs:49]: is the problem that `run_test` uses `panic` to report failures? the Swift runner has a couple custom error types we throw for unexpected failures, I wonder how much work it would be to switch to something like that so `run_test` could return a `Result`? otherwise I imagine this is going to make it really difficult to debug any failures we see in Evergreen

File: src/test/spec/unified_runner/operation.rs
Comment (kmahar) [inline src/test/spec/unified_runner/operation.rs:1997]: should we be using the test runner "matching" logic here or is this always an exact match?

**Review by kmahar — COMMENTED** [review]
looking good! few comments and questions

File: src/test/spec/unified_runner/operation.rs
Comment (abr-egn) [inline src/test/spec/unified_runner/operation.rs:80]: Is it useful for this to be a failure rather than an immediate `panic!`?

File: src/test/spec/unified_runner/operation.rs
Comment (abr-egn) [inline src/test/spec/unified_runner/operation.rs:2010]: Should there be handling for expected value but the entity is `None` / not `Bson`?

**Review by abr-egn — COMMENTED** [review]
Overall LG!  Just a few small comments beyond Kaitlin's.

File: src/test/atlas_planned_maintenance_testing/mod.rs
Comment (patrickfreed) [inline src/test/atlas_planned_maintenance_testing/mod.rs:75]: It might be better to first create a `PathBuf` from the environment variable here and then use `path.push("events.json")` to prevent any issues with Windows.

File: src/test/spec/unified_runner/operation.rs
Comment (patrickfreed) [inline src/test/spec/unified_runner/operation.rs:1989]: since we have a lot of nested matches and if lets, this might be more readable as one larger, flat match statement,
e.g.
```rust
match (result, operation.expectation) {
    // Have an entity result, expected one, need to check if equal
    (Ok(Entity::Bson(entity)), Expectation::Result { Some(expected_value), save_as_entity }) => { ... },

    // Got an ok response and expected one, but don't care about what it is, always success
    (Ok(_), Expectation::Result { None, _ }) => { ... }

    // Got an ok result, but expected an error, always a failure
    (Ok(entity), Expectation::Error(expected_error) => { ... }

    // ignore, always success
    (Ok(_), Expectation::Ignore) => { ... }

    // Got an error, but expected a result
    (Err(e), Expectation::Result { ... } | Expectation::Ignore) => { ... }

    // Got an error, expected an error
    (Err(e), Expectation::Error(expected_error)) => { ... }
}
```
To me this is a bit easier to reason about and also has the benefit of having the compiler verify that we've handled all cases properly. Also could combine the first two cases and keep a nested `if let`. Anyways, this is just my preference, feel free to leave it as-is if you think the current version is more readable.

File: src/test/atlas_planned_maintenance_testing/mod.rs
Comment (isabelatkinson) [inline src/test/atlas_planned_maintenance_testing/mod.rs:23]: this is a slightly more convenient way to get the name of the test executable (as opposed to parsing through `cargo` output). all of the tests in the crate are run via one executable that can be filtered by passing test names as arguments. `std::env::current_exe()` returns the name of the executable that's currently running, so when it's called within a test it'll return the name of that test executable. this test method then writes that name to a file that the workload executor can invoke via `$(cat exe_name.txt workload_executor)`. (I used `into_os_string` on the `PathBuf` returned rather than the `Debug` implementation because the latter includes quotations around the path that would need to be stripped out.)

File: src/test/atlas_planned_maintenance_testing/mod.rs
Comment (kmahar) [reply src/test/atlas_planned_maintenance_testing/mod.rs:65]: thanks for the context! 

in Swift, the top-level method `UnifiedTestRunner.runTests` is marked `throws`, so we can propagate errors all the way up to that level pretty easily, which leads the calling test case to throw an error and fail. 
for operations that expect errors we use a `do`/`catch` around the appropriate code, and for some unexpected errors we catch them and re-throw them with extra context added. some unexpected errors just get directly propagated up though, like anything during test setup I believe.

I suppose this all ends up being more straightforward and concise in Swift than Rust since the Swift compiler does not care about what type of error a method throws, so we don't have to worry about converting errors all to some consistent type. also, we can just `try` things that throw in Swift, but since we can't use `?` on async return values working with the results is a lot more verbose here. 🤔

File: src/event/sdam/mod.rs
Comment (abr-egn) [inline src/event/sdam/mod.rs:169]: Maybe this should go in some common place so it can be shared?

File: src/test/atlas_planned_maintenance_testing/mod.rs
Comment (abr-egn) [reply src/test/atlas_planned_maintenance_testing/mod.rs:72]: I think it's reasonable to take this approach here; it should work for what's needed and it keeps this PR from ballooning.

I do think there should be a bug to track converting the test code to return errors rather than using `panic!`.  I'm not sure how unwind-safe the test runner actually is; I don't think there are going to be memory-safety problems, but it seems quite possible the runner could end up with broken assumptions that cause flaky tests.  It seems roughly equivalent to the cancellation issue.

Comment (kmahar) [reply src/test/atlas_planned_maintenance_testing/mod.rs:72]: I agree with Abraham, deferring test runner refactoring to use errors to a follow up ticket SGTM

**Review by kmahar — COMMENTED** [review]
took another pass through - everything here LGTM, but I'll hold off on approving in case there end up being any more changes

**Review by isabelatkinson — COMMENTED** [review]
Moving this out of draft/WIP status as the planned maintenance tests are finally all passing. Going to hold off on fixing the merge conflicts until this is ready to merge so I can rebase down to one commit.

**Review by abr-egn — APPROVED** [review]

**Review by patrickfreed — APPROVED** [review]
LGTM! Awesome job with this.

Just looks like there are some merge conflicts and possibly an MSRV issue that need to be fixed up.

**Pull request merged**
